src/APIBasedAuth.ts

```diff
@@ -0,0 +1,192 @@
+import axios, {
+  AxiosError,
+  AxiosInstance,
+  AxiosRequestConfig,
+  AxiosResponse
+} from 'axios';
+
+/* istanbul ignore next */
+const formatAxiosError = <ErrorResponse = any>(
+  error: AxiosError<ErrorResponse> | unknown
+) => {
+  if (!axios.isAxiosError(error)) {
+    throw new AxiosError(String(error));
+  }
+  throw error.response?.data;
+};
+
+class APIBasedAuth {
+  private client: AxiosInstance;
+  private clientOptions: APIBasedAuth.Config;
+  private session?: APIBasedAuth.Session;
+  private tokenData?: APIBasedAuth.TokenData;
+
+  constructor(options: APIBasedAuth.Config) {
+    const required = ['baseURL', 'clientId'];
+    for (const param of required) {
+      // eslint-disable-next-line security/detect-object-injection
+      if (!options[param]) {
+        throw new Error(`APIBasedAuth ctor param ${param} is required`);
+      }
+    }
+
+    this.client = axios.create({
+      baseURL: options.baseURL,
+      withCredentials: options.withCredentials
+    });
+    this.client.interceptors.request.use(this._setAuthorizationHeader);
+
+    this.clientOptions = options;
+  }
+
+  /* istanbul ignore next */
+  private async _setAuthorizationHeader(
+    request: AxiosRequestConfig<unknown>
+  ): Promise<AxiosRequestConfig<unknown>> {
+    if (this.tokenData?.access_token) {
+      request.headers = {
+        ...request?.headers,
+        Authorization: `Bearer ${this.tokenData.access_token}`
+      };
+    }
+    return request;
+  }
+
+  public confirmPasswordlessAuth(
+    baseInput: Omit<APIBasedAuth.VerifyPasswordlessAuthData, 'clientId'>
+  ) {
+    const input = {
+      ...baseInput,
+      // if session or username is not present get stored properties
+      ...(!baseInput.session || !baseInput.username ? this.session : null)
```

Comment:
In the case that `session` needs to come from some other storage (user clicks login button and opens new tab for example) would the caller need to handle that explicitly with `baseInput`?